rust/src/daemon.rs

```diff
@@ -131,6 +149,167 @@ fn deployment_populate_variant_origin(
+            msg: format!("Unexpected message: {o:?}"),
+        },
+    };
+    write_message(&client, reply).context("Writing client reply")?;
+    tracing::debug!("Acknowleged client");
+    Ok(())
+}
+
+static SHUTDOWN_SIGNAL: Lazy<Mutex<Option<Sender<()>>>> = Lazy::new(|| Mutex::new(None));
+
+fn run_acknowledgement_worker(listener: OwnedSocketlike) {
+    tracing::debug!("Processing clients...");
+    loop {
+        let sock = match rustix::net::accept(&listener) {
+            Ok(s) => s,
+            Err(e) => {
+                tracing::warn!("Failed to accept client: {e}");
+                continue;
+            }
+        };
+        std::thread::spawn(move || {
+            if let Err(e) = client_hello(sock.into_socketlike(), Ok(())) {
+                tracing::warn!("error acknowledging client: {e}");
+            }
+        });
+    }
+}
+
+/// Ensure all asynchronous tasks in this Rust half of the daemon code are stopped.
+/// Called from C++.
+pub(crate) fn daemon_terminate() {
+    if let Some(chan) = (*SHUTDOWN_SIGNAL).lock().unwrap().take() {
+        let _ = chan.send(());
+    }
+}
+
+fn process_one_client(listener: OwnedSocketlike, e: anyhow::Error) -> Result<()> {
+    let incoming = rustix::net::accept(&listener)?;
+    // Send the error to the client
+    client_hello(incoming.into_socketlike(), Err(e))?;
+    // We've successfully sent the error; the caller of this function will
+    // propagate the error.
+    Ok(())
+}
+
+/// Perform initialization steps required by systemd service activation.
+///
+/// This ensures that the system is running under systemd, then receives the
+/// socket-FD for main IPC logic, and notifies systemd about ready-state.
+pub(crate) fn daemon_main(debug: bool) -> Result<()> {
+    let handle = tokio::runtime::Handle::current();
+    let _tokio_guard = handle.enter();
+    if !systemd::daemon::booted()? {
+        return Err(anyhow!("not running as a systemd service"));
+    }
+    let init_res: Result<()> = crate::ffi::daemon_init_inner(debug).map_err(|e| e.into());
+    tracing::debug!("Initialization result: {init_res:?}");
+
+    let mut fds = systemd::daemon::listen_fds(false)?.iter();
+    let (listener, init_res) = match fds.next() {
+        None => {
+            // If started directly via `systemctl start` or DBus activation, we
+            // directly propagate the error back to our exit code without even bothering
+            // with a socket.
+            init_res.context("Initializing via dbus")?;
+            tracing::debug!("Initializing directly (not socket activated)");
+            let listener = {
+                let socket = rustix::net::socket(
+                    rustix::net::AddressFamily::UNIX,
+                    rustix::net::SocketType::SEQPACKET,
+                    rustix::net::Protocol::from_raw(0),
+                )?;
+                let addr = crate::client::sockaddr()?;
+                rustix::net::sockopt::set_socket_reuseaddr(&socket, true)?;
```

Comment:
Upon a `start-daemon` invocation, the dameon gives a `error: Address already in use (os error 98)` . This happens when we try bind the socket to the address despite the `so-reuseaddr` option enabled

Reply:
OK right, we may need to drop the code that tries to bind to the socket in the daemon code.  

Hmm...I think what we really want is to make systemd *always* pass the socket to us even when we're started via dbus.

I think we may be able to do this via something like
```
[Service]
StandardInput=fd:rpm-ostreed.socket
```
perhaps?

Then we should always be able to find our socket on stdin, if that works.  Needs testing

Reply:
OK that didn't work in my testing, but using `Sockets=rpm-ostreed.socket` did seem to work.

I was actually testing with this:

```
[root@cosa-devsh ~]# cat /etc/systemd/system/foo.socket 
[Socket]
ListenStream=/run/foo.sock
SocketMode=0600

[Install]
WantedBy=sockets.target
[root@cosa-devsh ~]# cat /etc/systemd/system/foo.service 
[Service]
Sockets=foo.socket
Type=oneshot
ExecStart=env
ExecStart=ls -al /proc/self/fd/0
[root@cosa-devsh ~]# 
```

Using `Sockets=` I do see when I `systemctl start foo` that it has the expected `LISTEN_FDS=1` in the environment.

Reply:
For the record, I seem to have just been wrong here, I do seem to get the socket always passed.

I suspect the problem we have been seeing here is in the case where the socket unit isn't active.  That will happen in some development scenarios (e.g. copying new rpm-ostree binary, but not the systemd unit - or even more commonly, forgetting to systemctl daemon-reload etc.)

Arguably perhaps, the daemon should just yell loudly about this and the client can also loudly complain if the socket is missing, and just fall back to dbus.